scrapy/responsetypes.py

```diff
@@ -102,18 +117,45 @@ def from_body(self, body):
         else:
             return self.from_mimetype('text')
 
+    def _guess_content_type(self, content_type=None, content_disposition=None, url=None, filename=None):
+        if content_type:
+            return tuple(content_type)
+
+        if content_disposition:
+            filename = content_disposition.split(b';')[-1].split(b'=')[-1].strip(b'"\'').decode()
+
+        elif url:
+            filename = url
+
+        if filename:
+            return (self.mimetypes.guess_type(filename)[0].encode(),)
+
+        return None
+
     def from_args(self, headers=None, url=None, filename=None, body=None):
         """Guess the most appropriate Response class based on
         the given arguments."""
+        if not body:
+            body = b''
+
+        body = body[:RESOURCE_HEADER_BUFFER_LENGTH].replace(b"\x00", b"")
         cls = Response
-        if headers is not None:
-            cls = self.from_headers(headers)
-        if cls is Response and url is not None:
-            cls = self.from_filename(url)
-        if cls is Response and filename is not None:
-            cls = self.from_filename(filename)
-        if cls is Response and body is not None:
-            cls = self.from_body(body)
+        http_origin = True
+        content_types = None
+
+        if url and urlparse(url).scheme not in ("http", "https"):
+            http_origin = False
+
+        if headers:
+            content_types = self._guess_content_type(content_type=headers.getlist(b'Content-Type'),
+                                                  content_disposition=headers.get(b'Content-Disposition'),
+                                                  url=url, filename=filename)
+        else:
+            content_types = self._guess_content_type(url=url, filename=filename)
+
```

Comment:
If you move a bit more logic to `_guess_content_type`, you could reduce this part to 2 lines without losing readability, I think:

```suggestion
        http_origin = not url or urlparse(url).scheme in ("http", "https")
        content_types = self._guess_content_type(headers=headers, url=url, filename=filename)
```